sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/SingletonPatternCheck.kt

```diff
@@ -0,0 +1,131 @@
+        }
+    }
+}
+
+private class SingletonClassCandidateExtractor : KtTreeVisitor() {
+
+    val singletonClassCandidates = mutableSetOf<String>()
+
+    override fun visitClass(klass: KtClass) {
+        if (isSingletonClassCandidate(klass)) {
+            klass.fqName?.let {
+                singletonClassCandidates.add(it.toString())
+            }
+        }
+    }
+}
+
+private fun isSingletonClassCandidate(klass: KtClass): Boolean {
+    val constructors = klass.allConstructors
+    return klass.isPrivate() || (constructors.isNotEmpty() && constructors.all(KtModifierListOwner::isPrivate))
+}
+
+private fun KtModifierListOwner.isPrivate(): Boolean = visibilityModifier()?.text == "private"
+
+private class SingleConstructorCallExtractor(
+    private val singletonClassCandidates: MutableSet<String>,
+    private val bindingContext: BindingContext,
+) : KtTreeVisitor() {
+
+    val singleConstructorCallByClass: MutableMap<String, KtCallExpression> = mutableMapOf()
+
+    override fun visitCallExpression(expression: KtCallExpression) {
+        val fqName = getCalledConstructorClassFqNameOrNull(expression) ?: return
+        if (!singletonClassCandidates.contains(fqName)) return
+        if (singleConstructorCallByClass.put(fqName, expression) != null) {
+            singleConstructorCallByClass.remove(fqName)
+            singletonClassCandidates.remove(fqName)
+        }
+    }
+
+    private fun getCalledConstructorClassFqNameOrNull(expression: KtCallExpression): String? {
+        val calleeReference = expression.calleeExpression as? KtReferenceExpression ?: return null
+        val calledConstructor = bindingContext[BindingContext.REFERENCE_TARGET, calleeReference] as? ConstructorDescriptor
+        return calledConstructor?.constructedClass?.fqNameSafe?.toString()
+    }
+}
+
+private fun isInitializingCall(callExpression: KtCallExpression): Boolean {
+    val property = callExpression.parent as? KtProperty ?: return false
+    return isDeclaredInCompanionObject(property)
+}
+
+private fun isLazyInitializingCall(callExpression: KtCallExpression, bindingContext: BindingContext): Boolean {
+    val bodyExpression = callExpression.parent as? KtBlockExpression ?: return false
+    if (bodyExpression.statements.last() != callExpression) return false
+    val call = (((bodyExpression.parent as? KtFunctionLiteral)
+        ?.parent as? KtLambdaExpression)
+        ?.parent as? KtLambdaArgument)
+        ?.parent as? KtCallExpression ?: return false
+
+    val property = (call.parent as? KtPropertyDelegate)?.parent as? KtProperty ?: return false
+    if (call.valueArguments.size != 1 || !isDeclaredInCompanionObject(property)) return false
+    val calleeReference = call.calleeExpression as? KtReferenceExpression ?: return false
+    return bindingContext[BindingContext.REFERENCE_TARGET, calleeReference]?.name?.toString() == "lazy"
+}
+
+private fun isDeclaredInCompanionObject(property: KtProperty): Boolean =
```

Comment:
What about singleton instance fields declared on file level? Do we ignore those on purpose?

Reply:
Unfortunately yes, because a TLD instance field implies that the constructor is not private. But thinking about it again: if the constructor is not private, but the class is, we could do the check also for TLD instance fields. I'll add that.